### DISCUSSION TITLE
制作的整体思路讨论

### DISCUSSION THREAD

**Discussion by gongfuture** [XDTechno/XDLiveSim#1] [General]
QwQ @AlseFum

**Comment by AlseFum** [comment]
我的想法现在也不太明确，大体上是希望：
1.行动的个体称为单元。单元行动的空间称为场景。
2.单元内部有能用于控制单元行动的逻辑数据，这个逻辑数据既能从亲代遗传到子代，又能随机发生变异或者交换
3.单元对外有各种行动，能对外界的刺激（其他单元的行动）做出反应
4.场景是二维网格型的，有多种地形设置。还有一些实体。实体只能被动地作出反应而且不能改变行为逻辑（可能可以移动）
5.单元的行动是按时间轴离散地来的
然后具体实现，我的技术栈在HTML和js方面，有vue，但是没有react

感觉得先把实现的思路整出来

**Reply by ligeaaa** [reply]
2.单元内部有能用于控制单元行动的逻辑数据，这个逻辑数据既能从亲代遗传到子代，又能随机发生变异或者交换
这个让我想到了遗传算法（）就是一个模拟dna变异交换的一个过程的一个算法，用来找非线性最优解的。但是可能要对“逻辑数据”做一个具体的定义，也就是定义好基因，有哪些基因段，这些基因段中又有哪些成分。当然，这个算法效率不高）））我也没很多了解，权当抛砖引玉

**Reply by AlseFum** [reply]
我当初也看到了这个
问题是这个逻辑是怎么控制行为的，又是怎么改变的

我之前有一个想法是整一串字节码，逐字读取，发现匹配的有意义的字符段就运行，没意义就跳过。到最后或者有别的什么条件就结束这条逻辑的运行。
然后随机就是改变无意义的字符段使其可能有意义。

进阶一点的可能就是从逻辑字符段构造一个状态机运行，这样效率高一点。逻辑改变之后就更新。

**Reply by Xinweigh** [reply]
>我之前有一个想法是整一串字节码，逐字读取，发现匹配的有意义的字符段就运行，没意义就跳过。到最后或者有别的什么条件就结束这条逻辑的运行 

这里的字节码其实可以用整型值来代替，整型值对应行为，大概可以简化编码。

那么基于这个思路，考虑面向对象的话，各个种类的生命体可以作为对象，其内部定义一个生命周期的操作（诞生、生长、繁殖和遗传、变异、死亡和凋亡等，具体要考虑多少种需要列个清单）
实际生命活动由上述的字节码或整型数组（DNA）控制，难点在于设计每个有意义的操作对应的执行行为、执行后的状态转移以及变异后的运行。执行行为要求适配该种生物，状态转移要求转移到该个体的DNA定义的行为中去，变异的淘汰可以定义为缺失关键生命活动或执行行为冲突致使个体死亡。

**Reply by AlseFum** [reply]
>  这里的字节码其实可以用整型值来代替，整型值对应行为，大概可以简化编码。
> 
> 那么基于这个思路，考虑面向对象的话，各个种类的生命体可以作为对象，其内部定义一个生命周期的操作（诞生、生长、繁殖和遗传、变异、死亡和凋亡等，具体要考虑多少种需要列个清单） 实际生命活动由上述的字节码或整型数组（DNA）控制，难点在于设计每个有意义的操作对应的执行行为、执行后的状态转移以及变异后的运行。执行行为要求适配该种生物，状态转移要求转移到该个体的DNA定义的行为中去，变异的淘汰可以定义为缺失关键生命活动或执行行为冲突致使个体死亡。

就是整型值呀

所以现在的问题是生命体的行动有哪些

抽象一点，一个生命体需要三部分：输入（感知），记忆（逻辑和自身状态），输出（行动）。
可以认为输出一个特定的token这个生命体就死亡（或者说凋亡）了。
但是其实应该是生命体内部有个什么东西决定其是否死亡，跟它的逻辑无关

因此问题转到了生命体活动的场地上。

**Reply by ligeaaa** [reply]
不妨一开始设想的简单一点，一个生物，只有三个行为：1、行动，2、交配，3、死亡。每个行动都由具体的基因控制。
将生物抽象为一个点，世界抽象为一个二维有边界世界，生物相遇就会交配，交配后会在原地留下新生物，且10min后两个父母生物就会死亡。
接下来是行动的逻辑，处于简化考虑，生物只有上下左右移动x格的行动方案，每个时间段会按照基因做出响应行动。例如一个基因的文字表述是这样子的：该生物会在每个时间段随机选择移动方向，且移动x格。其中x由具体的基因片段A控制，A中有a1，a2，a3..ai，b1，b2，b3..bi等具体基因。每个生物交配的时候，参考基因交配和变异去将两个 生物的A基因片段进行组合，生成一个新的A基因片段，也就是他们孩子的走路逻辑

至于更复杂的行动，，，，

**Reply by AlseFum** [reply]
你这不是把逻辑都定好多了
还是得让逻辑决定除了死亡之外的行动

先确定一下场景是什么样的吧，以及生命体移动的顺序（？

**Reply by ligeaaa** [reply]
个人倾向于做一个2D有边界棋盘，这样子会比较简单（？）移动顺序之类的，可以直接回合制（？），然后在基因串里放入一个特定token用来决定每个回合的移动顺序

**Reply by gongfuture** [reply]
> 不妨一开始设想的简单一点，一个生物，只有三个行为：1、行动，2、交配，3、死亡。每个行动都由具体的基因控制。 将生物抽象为一个点，世界抽象为一个二维有边界世界，生物相遇就会交配，交配后会在原地留下新生物，且10min后两个父母生物就会死亡。 接下来是行动的逻辑，处于简化考虑，生物只有上下左右移动x格的行动方案，每个时间段会按照基因做出响应行动。例如一个基因的文字表述是这样子的：该生物会在每个时间段随机选择移动方向，且移动x格。其中x由具体的基因片段A控制，A中有a1，a2，a3..ai，b1，b2，b3..bi等具体基因。每个生物交配的时候，参考基因交配和变异去将两个 生物的A基因片段进行组合，生成一个新的A基因片段，也就是他们孩子的走路逻辑
> 
> 至于更复杂的行动，，，，

不考虑无性繁殖的吗（），我以为咱们是以细胞为基础进行模拟的

**Reply by aiyooooooo** [reply]
这么来看的话单元是可以移动的算是一个共识了？

**Reply by gongfuture** [reply]
> 这么来看的话单元是可以移动的算是一个共识了？

大概是的？

**Reply by Xinweigh** [reply]
[Designation.md](https://github.com/XDTechno/XDLiveSim/files/15451470/Designation.md)
工位摸鱼写了一些，大家看看（=°ω°)=

**Reply by gongfuture** [reply]
> [Designation.md](https://github.com/XDTechno/XDLiveSim/files/15451470/Designation.md) 工位摸鱼写了一些，大家看看（=°ω°)=

肥哥牛哇！看了一遍，我个人感觉没什么缺的了，现在的工作应该就是写一个生物通用的模板？大家看看还有没有其他要补充的？
~顺便，应该用什么技术栈搭建呢~

**Reply by Xinweigh** [reply]
> > [Designation.md](https://github.com/XDTechno/XDLiveSim/files/15451470/Designation.md) 工位摸鱼写了一些，大家看看（=°ω°)=
> 
> 肥哥牛哇！看了一遍，我个人感觉没什么缺的了，现在的工作应该就是写一个生物通用的模板？大家看看还有没有其他要补充的？ ~顺便，应该用什么技术栈搭建呢~

感谢认可( ´∀`)
技术栈考虑大家的实际情况吧，我这边是cpp和py（其他也可以学）
另外一点是建议对多线程甚至cuda支持良好，否则游戏进行到后期可能就会暴击CPU0（

**Reply by AlseFum** [reply]
好好好，这下跨平台力
我的技术主要堆在JS上，C++和Python也能写，但是不怎么会配置
感觉现在有两个方向
- 做出能用的图形化界面出来，要能看到这个生命体的活动，不一定需要实际运行
- 初步定下场景和生命体的标准，以便跨平台（？

关于场景和时间：
现在先简单做出N×N的矩形无界场景吧，不限制移动
时间可以做成一帧一帧，六十帧一刻的形式。生命体的行动在一开始决定好，在一定帧数之后实现。
如果有冲突的行动那么先动的生命体成功，后动的生命体失败。同时进行就随机。

然后现在的生命体首要的功能是移动和感知，以及最基本的吸收。反应（reaction）往后稍稍。
另外我认为生命体内部有必要存储几个属性，比如敏捷啊力量啊什么的（

**Reply by gongfuture** [reply]
坏了，看来只有我最菜了，只会一点点Python，C快忘完了，最近看了一点点Rust和TypeScript，只有自己写过的一点点小项目经历，没有大型项目经验，那我就专注GitHub维护吧（bushi）

**Reply by ligeaaa** [reply]
我技术栈主要在java（），虽然现在是在一家公司算法实习，用py比较多）但是感觉py很难说上什么技术栈）））总之前端以外的事情也许多多少少都能干点吧

**Reply by ligeaaa** [reply]
> > > [Designation.md](https://github.com/XDTechno/XDLiveSim/files/15451470/Designation.md) 工位摸鱼写了一些，大家看看（=°ω°)=
> > 
> > 
> > 肥哥牛哇！看了一遍，我个人感觉没什么缺的了，现在的工作应该就是写一个生物通用的模板？大家看看还有没有其他要补充的？ ~顺便，应该用什么技术栈搭建呢~
> 
> 感谢认可( ´∀`) 技术栈考虑大家的实际情况吧，我这边是cpp和py（其他也可以学） 另外一点是建议对多线程甚至cuda支持良好，否则游戏进行到后期可能就会暴击CPU0（

快上传到readme口牙

**Comment by AlseFum** [comment]
怎么没有人说话了？
还是得开个discord讨论加吹水才好；-）

**Reply by gongfuture** [reply]
我在忙期末考( ´ー`)起码得考完了这边才有空

**Reply by ligeaaa** [reply]
> 怎么没有人说话了？ 还是得开个discord讨论加吹水才好；-）

怎么已经没动静了！快把discord端上来吧！

**Reply by gongfuture** [reply]
> > 怎么没有人说话了？ 还是得开个discord讨论加吹水才好；-）
> 
> 怎么已经没动静了！快把discord端上来吧！

阿巴阿巴

**Reply by gongfuture** [reply]
https://discord.gg/EMkAc4YVXE
